internal/v4/auth_test.go

```diff
@@ -820,6 +874,118 @@ func (s *authSuite) TestDelegatableMacaroonWithBasicAuth(c *gc.C) {
+	c.Assert(err, jc.ErrorIsNil)
+
+	err = s.store.SetPerms(&id1.URL, "read", "bob")
+	c.Assert(err, jc.ErrorIsNil)
+
+	err = s.store.SetPerms(&id2.URL, "read", "bob")
+	c.Assert(err, jc.ErrorIsNil)
+
+	client := httpbakery.NewHTTPClient()
+
+	now := time.Now()
+	var gotBody json.RawMessage
+	httptesting.AssertJSONCall(c, httptesting.JSONCallParams{
+		Handler:  s.srv,
+		URL:      storeURL("delegatable-macaroon"),
+		JSONBody: &id2.URL,
+		ExpectBody: httptesting.BodyAsserter(func(c *gc.C, m json.RawMessage) {
+			gotBody = m
+		}),
+		Do:           bakeryDo(nil),
+		ExpectStatus: http.StatusOK,
+	})
+
+	c.Assert(gotBody, gc.NotNil)
+	var m macaroon.Macaroon
+	err = json.Unmarshal(gotBody, &m)
+	c.Assert(err, gc.IsNil)
+
+	caveats := m.Caveats()
+	foundExpiry := false
+	for _, cav := range caveats {
+		cond, arg, err := checkers.ParseCaveat(cav.Id)
+		c.Assert(err, gc.IsNil)
+		switch cond {
+		case checkers.CondTimeBefore:
+			t, err := time.Parse(time.RFC3339Nano, arg)
+			c.Assert(err, gc.IsNil)
+			c.Assert(t, jc.TimeBetween(now.Add(v4.DelegatableMacaroonExpiry), now.Add(v4.DelegatableMacaroonExpiry+time.Second)))
+			foundExpiry = true
+		}
+	}
+	c.Assert(foundExpiry, jc.IsTrue)
+
+	// Now check that we can use the obtained macaroon to do stuff
+	// as the declared user.
+	s.discharge = func(_, _ string) ([]checkers.Caveat, error) {
+		return nil, errgo.New("no discharge")
+	}
+	s.dischargeTerms = func(_, _ string) ([]checkers.Caveat, error) {
+		return nil, errgo.New("no discharge")
+	}
+
+	// First check that we require authorization to access the charm.
+	rec := httptesting.DoRequest(c, httptesting.DoRequestParams{
+		Handler: s.srv,
+		URL:     storeURL("~charmers/precise/terms/meta/id-name"),
+		Method:  "GET",
+	})
+	c.Assert(rec.Code, gc.Equals, http.StatusProxyAuthRequired)
+
+	// Then check that the request succeeds if we provide the delegatable
+	// macaroon.
+
+	client = httpbakery.NewHTTPClient()
+	u, err := url.Parse("http://127.0.0.1")
+	c.Assert(err, gc.IsNil)
+	err = httpbakery.SetCookie(client.Jar, u, macaroon.Slice{&m})
+	c.Assert(err, gc.IsNil)
+
+	// Test access to metadata.
+	httptesting.AssertJSONCall(c, httptesting.JSONCallParams{
+		Handler: s.srv,
+		URL:     storeURL("~charmers/precise/terms/meta/id-name"),
+		ExpectBody: params.IdNameResponse{
+			Name: "terms",
+		},
+
+		ExpectStatus: http.StatusOK,
+		Do:           bakeryDo(client),
+	})
+
+	// Test access to archive.
+	rec = httptesting.DoRequest(c, httptesting.DoRequestParams{
+		Handler: s.srv,
+		URL:     storeURL("~charmers/precise/terms/archive"),
+		Do:      bakeryDo(client),
+	})
+	c.Assert(rec.Code, gc.Equals, http.StatusOK)
+
+	// Test access to archive of another charm.
+	httptesting.DoRequest(c, httptesting.DoRequestParams{
+		Handler:     s.srv,
+		URL:         storeURL("~charmers/precise/terms-10/archive"),
+		Do:          bakeryDo(client),
+		ExpectError: ".*third party refused discharge: cannot discharge: no discharge",
+	})
+}
+
```

Comment:
Again, I might have missed them, but I don't see any tests for the usual authorization (not the delegatable macaroon) with charms that have terms.